cpu/samd21/periph/adc.c

```diff
@@ -0,0 +1,361 @@
+
+int adc_sample(adc_t dev, int channel)
+{
+    int result = 0;
+    Adc *adc = 0;
+    switch (dev)
+    {
+    #if ADC_0_EN
+        case ADC_0:
+            adc = ADC_0_DEV;
+            /* Starts the ADC conversion */
+            while(adc_syncing(adc));
+            adc->SWTRIG.reg |= ADC_SWTRIG_START;
+        break;
+    #endif
+        default:
+            DEBUG("case default\n");
+            return -1;
+    }
+    /* MUST NOT BLOCK */
+    while(!(adc_get_status(adc) & ADC_0_STATUS_RESULT_READY))
+    {
+        DEBUG("STATUS NOT READY\n");
+    }
+    while(adc_syncing(adc));
+    result = adc->RESULT.reg;
+    /*CLEAR RESET FLAG */
+    adc_clear_status(adc, ADC_0_STATUS_RESULT_READY);
+    if (adc_get_status(adc) & ADC_0_STATUS_OVERRUN) 
+    {
+        adc_clear_status(adc, ADC_0_STATUS_OVERRUN);
+        DEBUG("OVERRUN\n");
+        return -1;
+    }
+    return result;
+}
+
+void adc_poweron(adc_t dev) 
+{
+    switch (dev) 
+    {
+    #if ADC_0_EN
+        case ADC_0:
+            /* Setup generic clock mask for adc */
+            PM->APBCMASK.reg |= PM_APBCMASK_ADC;
+            break;
+    #endif
+        default:
+            break;    
+    }
+}
+
+void adc_poweroff(adc_t dev)
+{
+    Adc *adc = 0;
+    
+    switch (dev) 
+    {
+    #if ADC_0_EN
+        case ADC_0:
+            adc = ADC_0_DEV;
+            while(adc_syncing(adc));
+            /* Disable */
+            adc->CTRLA.reg &= ~ADC_CTRLA_ENABLE;
+            break;
+    #endif
+        default:
+            break;
+    }
+}
+
+int adc_map(adc_t dev, int value, int min, int max)
+{
+    DEBUG("adc_map Not implemented!\n");
+    return 0;
+}
+
+float adc_mapf(adc_t dev, int value, float min, float max)
+{
+    DEBUG("adc_mapf Not implemented!\n");
+    return 0.0;
+}
+
+bool adc_syncing(Adc* adc)
+{
+    if(adc->STATUS.reg & ADC_STATUS_SYNCBUSY){
+        return true;
+    }
+    return false;
+}
+
+/* Configure ADC with defined Resolution */
+int adc_configure_with_resolution(Adc* adc, uint32_t precision)
+{
+    adc_poweron(ADC_0);
+    uint8_t  divideResult = ADC_0_DIV_RES_DEFAULT;
+    uint32_t resolution = ADC_0_RES_16BIT;
+    uint32_t accumulate = ADC_0_ACCUM_DEFAULT;
+    if(adc->CTRLA.reg & ADC_CTRLA_SWRST)
+        return -1;
+    if(adc->CTRLA.reg & ADC_CTRLA_ENABLE)
+        return -1;
+    
+    /* GCLK Setup*/
+    GCLK->CLKCTRL.reg = (uint32_t)((GCLK_CLKCTRL_CLKEN
+                          | GCLK_CLKCTRL_GEN_GCLK0 
+                          | (ADC_GCLK_ID << GCLK_CLKCTRL_ID_Pos)));
+
+     /* Pin Muxing */
+    ADC_0_PORT.PINCFG[ ADC_0_POS_INPUT ].bit.PMUXEN = 1;
+    ADC_0_PORT.PMUX[ ADC_0_POS_INPUT / 2].bit.PMUXE = 1;
+
+    /* ADC_0_POS_INPUT Input */
+    ADC_0_PORT.DIRCLR.reg = (1 << ADC_0_POS_INPUT);
+    ADC_0_PORT.PINCFG[ADC_0_POS_INPUT].bit.INEN = true;
+    ADC_0_PORT.PINCFG[ADC_0_POS_INPUT].bit.PULLEN = false; 
+
+    if(ADC_0_NEG_INPUT != ADC_INPUTCTRL_MUXNEG_GND)
+    {
+        ADC_0_PORT.DIRCLR.reg = (1 << ADC_0_NEG_INPUT);
```

Comment:
According to the periph_conf, `ADC_0_POS_INPUT = ADC_INPUTCTRL_MUXNEG_GND = ADC_INPUTCTRL_MUXNEG_GND_Val  << ADC_INPUTCTRL_MUXNEG_Pos `. With  `ADC_INPUTCTRL_MUXNEG_GND_Val = 0x18` and `ADC_INPUTCTRL_MUXNEG_Pos = 8` you won't get a leftshift by 18 what I expected